Add rubocop

marcheing opened the pull request:
This adds the [rubocop](https://github.com/bbatsov/rubocop) gem, a configuration based on the [standards](https://github.com/mezuro/prezento/wiki/Standards) of the project and a rake task to rub rubocop on cli and on Travis.

dread-uo commented:
Shouldn't the refactoring part be on a different PR?

marcheing commented:
The refactor was needed to eliminate all of rubocop's warnings

dread-uo commented:
Ok, it's fine for me  :smiley_cat: 
Is everyone else satisfied with those standards?

rafamanzo commented on app/models/validators/range_overlapping_validator.rb:
```diff
@@ -12,6 +12,6 @@ def overlaps?(range1, range2)
     return true if range1.beginning >= range2.beginning && range1.end <= range2.end
     return true if range1.beginning <= range2.beginning && range1.end >= range2.end
-    return false
+    false
```
I'd ignore the complaint for this specific case

rafamanzo commented on features/support/env.rb:
```diff
@@ -41,7 +41,7 @@
   DatabaseCleaner.strategy = :transaction
 rescue NameError
-  raise "You need to add database_cleaner to your Gemfile (in the :test group) if you wish to use it."
+  raise 'You need to add database_cleaner to your Gemfile (in the :test group) if you wish to use it.'
```
This is `cucumber-rails` generated code. I'd ignore the warnings for this case.

marcheing replied:
Not entirely. This is mostly the generated code but some modifications were made. Adding to the ignore list anyway on 4b853f1bb7735e3cb8ebd895327f5a9d107bf8ec

rafamanzo commented on spec/controllers/kalibro_ranges_controller_spec.rb:
```diff
@@ -1,13 +1,13 @@
 require 'rails_helper'
 
-RSpec.describe KalibroRangesController, :type => :controller do
+RSpec.describe KalibroRangesController, type: :controller do
```
We could just remove the `type` and RSpec will infer by itself it.

rafamanzo commented on spec/controllers/metric_configurations_controller_spec.rb:
```diff
@@ -1,18 +1,20 @@
 require 'rails_helper'
 
-RSpec.describe MetricConfigurationsController, :type => :controller do
+RSpec.describe MetricConfigurationsController, type: :controller do
```
We could just remove the type and RSpec will infer by itself it.

rafamanzo commented on spec/controllers/metric_snapshots_controller_spec.rb:
```diff
@@ -1,6 +1,6 @@
 require 'rails_helper'
 
-RSpec.describe MetricSnapshotsController, :type => :controller do
+RSpec.describe MetricSnapshotsController, type: :controller do
```
We could just remove the type and RSpec will infer by itself it.

rafamanzo commented on spec/controllers/tests_controller_spec.rb:
```diff
@@ -22,7 +22,7 @@
 end
 
-RSpec.describe TestsController, :type => :controller do
+RSpec.describe TestsController, type: :controller do
```
We could just remove the type and RSpec will infer by itself it.

rafamanzo commented on spec/models/kalibro_configuration_spec.rb:
```diff
@@ -1,6 +1,6 @@
 require 'rails_helper'
 
-RSpec.describe KalibroConfiguration, :type => :model do
+RSpec.describe KalibroConfiguration, type: :model do
```
We could just remove the type and RSpec will infer by itself it.

rafamanzo commented on spec/models/kalibro_range_spec.rb:
```diff
@@ -1,8 +1,7 @@
 
-RSpec.describe KalibroRange, :type => :model do
-
+RSpec.describe KalibroRange, type: :model do
```
We could just remove the type and RSpec will infer by itself it.

rafamanzo commented on spec/models/metric_configuration_spec.rb:
```diff
@@ -1,6 +1,6 @@
 require 'rails_helper'
 
-RSpec.describe MetricConfiguration, :type => :model do
+RSpec.describe MetricConfiguration, type: :model do
```
We could just remove the type and RSpec will infer by itself it.

rafamanzo commented on spec/models/metric_snapshot_spec.rb:
```diff
@@ -1,6 +1,6 @@
 require 'rails_helper'
 
-RSpec.describe MetricSnapshot, :type => :model do
+RSpec.describe MetricSnapshot, type: :model do
```
We could just remove the type and RSpec will infer by itself it.

rafamanzo commented on spec/models/reading_group_spec.rb:
```diff
@@ -1,6 +1,6 @@
 require 'rails_helper'
 
-RSpec.describe ReadingGroup, :type => :model do
+RSpec.describe ReadingGroup, type: :model do
```
We could just remove the type and RSpec will infer by itself it.

rafamanzo commented on spec/models/reading_spec.rb:
```diff
@@ -1,6 +1,6 @@
 require 'rails_helper'
 
-RSpec.describe Reading, :type => :model do
+RSpec.describe Reading, type: :model do
```
We could just remove the type and RSpec will infer by itself it.

rafamanzo commented on spec/models/validators/color_validator_spec.rb:
```diff
@@ -1,27 +1,27 @@
 require 'validators/color_validator'
 
-RSpec.describe ColorValidator, :type => :model do
+RSpec.describe ColorValidator, type: :model do
```
We could just remove the type and RSpec will infer by itself it.

rafamanzo commented on spec/models/validators/interval_validator_spec.rb:
```diff
@@ -1,35 +1,35 @@
 require 'validators/interval_validator'
 
-RSpec.describe IntervalValidator, :type => :model do
+RSpec.describe IntervalValidator, type: :model do
```
We could just remove the type and RSpec will infer by itself it.

rafamanzo commented on spec/models/validators/range_overlapping_validation_spec.rb:
```diff
@@ -1,7 +1,7 @@
 require 'validators/range_overlapping_validator'
 
-RSpec.describe RangeOverlappingValidator, :type => :model do
+RSpec.describe RangeOverlappingValidator, type: :model do
```
We could just remove the type and RSpec will infer by itself it.

rafamanzo commented on spec/routing/kalibro_configurations_routing_spec.rb:
```diff
@@ -1,43 +1,41 @@
-
-    it "routes to #all" do
-      expect(:get => "/kalibro_configurations").to route_to("kalibro_configurations#all")
+RSpec.describe KalibroConfigurationsController, type: :routing do
```
We could just remove the type and RSpec will infer by itself it.

rafamanzo commented on spec/routing/kalibro_ranges_routing_spec.rb:
```diff
@@ -1,30 +1,29 @@
-
-    it "routes to #index" do
-      expect(:get => "/metric_configurations/3/kalibro_ranges").to route_to("kalibro_ranges#index", :metric_configuration_id => "3")
+RSpec.describe KalibroRangesController, type: :routing do
```
We could just remove the type and RSpec will infer by itself it.

rafamanzo commented on spec/routing/metric_configurations_routing_spec.rb:
```diff
@@ -1,26 +1,25 @@
-
-    it "routes to #create" do
-      expect(:post => "/metric_configurations").to route_to("metric_configurations#create")
+RSpec.describe MetricConfigurationsController, type: :routing do
```
We could just remove the type and RSpec will infer by itself it.

rafamanzo commented on spec/routing/metric_snapshots_routing_spec.rb:
```diff
@@ -1,38 +1,37 @@
-
-    it "routes to #index" do
-      expect(:get => "/metric_snapshots").to route_to("metric_snapshots#index")
+RSpec.describe MetricSnapshotsController, type: :routing do
```
We could just remove the type and RSpec will infer by itself it.

rafamanzo commented on spec/routing/reading_groups_routing_spec.rb:
```diff
@@ -1,31 +1,29 @@
-
-    it "routes to #all" do
-      expect(:get => "/reading_groups").to route_to("reading_groups#all")
+RSpec.describe ReadingGroupsController, type: :routing do
```
We could just remove the type and RSpec will infer by itself it.

rafamanzo commented on spec/routing/readings_routing_spec.rb:
```diff
@@ -1,31 +1,29 @@
-
-    it "routes to #index" do
-      expect(:get => "reading_groups/1/readings").to route_to("readings#index", :reading_group_id => "1")
+RSpec.describe ReadingsController, type: :routing do
```
We could just remove the type and RSpec will infer by itself it.

rafamanzo commented on spec/routing/tests_routing_spec.rb:
```diff
@@ -1,10 +1,12 @@
 
-describe TestsController, :type => :routing do
-  describe "routing" do
+describe TestsController, type: :routing do
```
We could just remove the type and RSpec will infer by itself it.

rafamanzo commented:
What do you think about removing it from Gemfile and adding installation instructions to the README? This way nobody is compelled to install it if is not willing to use it.

What do you think about disabling complaints about begin/rescue synthax? Is there any strong argument about it? If yes and not, what about restoring the changes made to begin/recues on the code?

The same as above for ifs?

Could you provide more details on the reasons for the changes at `lib/tasks/cucumber.rake`? Specially the new `begin` statement intrigued me.

I've added several specific comments on commit https://github.com/mezuro/kalibro_configurations/pull/79/commits/bf0ca9fe68fcf50e351de51d6920bfe1c98c1241

rafamanzo commented on app/models/validators/range_overlapping_validator.rb:
```diff
@@ -8,10 +8,30 @@ def validate(record)
+
+  private
+
+  # Range 1 - |------| <>    |------|
```
Nice comments! Such illustrativiness :dog:

rafamanzo commented on app/controllers/application_controller.rb:
```diff
@@ -2,4 +2,10 @@ class ApplicationController < ActionController::Base
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :null_session
+
+  def respond_with_json(data, status)
```
Could not we give more use to this method than just `app/controllers/metric_configurations_controller.rb`? What do you think about an issue mapping this? Looks like we can remove lots of duplication.

marcheing replied:
:ok:, an issue will be added after this PR.

diegoamc commented on app/controllers/metric_configurations_controller.rb:
```diff
@@ -7,51 +6,37 @@ def exists
-      respond_to do |format|
-        format.json { render json: {errors: @metric_configuration.errors.full_messages + metric_snapshot.errors.full_messages}, status: :unprocessable_entity }
-      end
+      data = { errors: @metric_configuration.errors.full_messages + metric_snapshot.errors.full_messages }
```
This line is exactly the same as L16. Can't we avoid this duplication?

marcheing replied:
They're both the same line of code but may be run on different preconditions because both `valid_metric_snapshot_for_configuration` and `save_metric_configuration_with_snapshot` may change the errors ([here](https://github.com/mezuro/kalibro_configurations/blob/add_rubocop/app/models/metric_configuration.rb#L33) and [here](https://github.com/mezuro/kalibro_configurations/blob/add_rubocop/app/controllers/metric_configurations_controller.rb#L78)), so they must be collected after the changes occur;
Maybe it can be done with a larger refactor, but I was trying to avoid that on this PR, only doing enough to make rubocop happy.

diegoamc commented on app/controllers/reading_groups_controller.rb:
```diff
@@ -56,19 +56,17 @@ def destroy
+  def set_reading_group
+    @reading_group = ReadingGroup.find(params[:id].to_i)
+    true
+  rescue ActiveRecord::RecordNotFound => exception
```
I'm seeing a lot of this `rescue` block. What do you think if we turned it into a `rescue_from` clause on `ApplicationController`?

marcheing replied:
What about after this PR? This looks like it should be handled along with the `respond_with_json`refactors on a single PR.

diegoamc commented on app/models/validators/range_overlapping_validator.rb:
```diff
@@ -8,10 +8,30 @@ def validate(record)
+
+  private
+
+  # Range 1 - |------| <>    |------|
```
:clap:

diegoamc commented:
Another thing we could do is to define a `code_style` group in our Gemfile, making installation explict if you want to use rubocop. Would be something like:
```ruby
group :code_style do
  gem 'rubocop', '>= 0.36'
end
```

Caveat: I believe, if rubocop is not wanted, the first time one install gems after this suggestion (if it is agreed on) will have to be with the command

```shell
bundle install --without code_style
```

After that, bundler will remember the last chosen installation options and explicitly leaving out the `code_style` group will not be needed.

What do you think?

marcheing commented:
If the gem is gonna be optional, then it's better just to not have it on gemfile at all. The instructions were added to the readme, so if anyone is concerned about keeping the style, then they just have to download rubocop and run it by themselves.

marcheing commented:
@rafamanzo I've removed the gem and added the instructions to the readme.
There was no new begin statement added to the `cucumber.rake` file, it simply added two spaces after the unless on each line.
I really prefer the removal of the [redundant begin](https://github.com/bbatsov/ruby-style-guide#begin-implicit) from the methods, in my opinion it doesn't damage the readability at all.
On the other hand, I can't seem to find any strong argument about the conditional assignment, so I can agree to remove that one. On the other hand, I know @danielkza likes this kind of assignment, so can we get a vote for these ones?

rafamanzo commented:
Regarding the redundant begin I can ague that it actually leads new users to error by adding new logic before the `rescue` and never knowing that she/he is supposed to add a `begin` in order to avoid this new logic from triggering unexpected rescues. Despite being redundant, is there anything else that bothers you about those `begin`s? Because otherwise I'd keep them for clarity and keeping easy to new developers.

@dread-uo, @diegoamc and @otaviocv what is your opinion on **conditional assignment** removal? We are **pros**: @rafamanzo and **cons**: @marcheing, @danielkza.

rafamanzo commented:
@marcheing it have come to my attention now that a lot of auto generated code at `lib/tasks/cucumber.rake` was modified. Is this necessary? This may make our life hard when trying to update this file.

rafamanzo commented:
And thanks for the fixes in the latest code :trophy:

rafamanzo commented:
This brings a new question: how will we keep the code rubocop compliant if we never install it or run on travis? It'll certainly become a hell (I may be changing my mind about default installing it).

dread-uo commented:
I agree to remove the conditional assignments.

@rafamanzo, I think the easier way to ensure code compliance would be using rubocop plugins within your favorite editor, as @marcheing does on vim. Running it separately every time would rapidly become a nuisance.

diegoamc commented:
I agree to remove the conditional assignments.

We can program the `rake spec` task to run rubocop automatically too.

About @rafamanzo's question on keeping the rubocop compliant, if we create a group on Gemfile we can make travis install and run it, without having to install it on local machines.

rafamanzo commented:
Looks like the last commit backfired :crying_cat_face: 

Maybe it has not respected the configurations set?

marcheing commented:
Because of the changed lines, [this line](https://github.com/mezuro/kalibro_configurations/blob/add_rubocop/app/models/kalibro_range.rb#L23) is not being covered by the tests, which was previously [this line](https://github.com/mezuro/kalibro_configurations/blob/master/app/models/kalibro_range.rb#L18).
Ignore the blame btw ( ͡° ͜ʖ ͡°)
Another issue will be opened then ^^

rafamanzo commented:
Don't forget about a changelog entry :)

rafamanzo commented:
BTW, can we configure rubocop to check for changelog entries?

marcheing commented:
@rafamanzo No

diegoamc commented:
I've already lost track of what is missing on this PR. Everything is fixed now? Have we decided on how to keep the code rubocop compliant?

marcheing commented:
@diegoamc Everything is fixed and there is a rake task to run rubocop now, along with the travis script.
:tangerine:

diegoamc commented:
Thanks again for the work, @marcheing! :clap:
